pkg/net/socketpair_unix.go

```diff
@@ -22,36 +22,43 @@ import (
 	"fmt"
 	"net"
 	"os"
+	"syscall"
 
-	syscall "golang.org/x/sys/unix"
+	"golang.org/x/sys/unix"
 )
 
-const (
-	local = 0
-	peer  = 1
-)
-
-// SocketPair contains the file descriptors of a connected pair of sockets.
-type SocketPair [2]int
+// SocketPair contains the os.File of a connected pair of sockets.
+type SocketPair struct {
+	local, peer *os.File
+}
 
 // NewSocketPair returns a connected pair of sockets.
 func NewSocketPair() (SocketPair, error) {
-	fds, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
+	syscall.ForkLock.RLock()
+	defer syscall.ForkLock.RUnlock()
+	fds, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_STREAM, 0)
 	if err != nil {
-		return [2]int{-1, -1}, fmt.Errorf("failed to create socketpair: %w", err)
+		return SocketPair{nil, nil}, fmt.Errorf("failed to create socketpair: %w", err)
 	}
+	unix.CloseOnExec(fds[0])
+	unix.CloseOnExec(fds[1])
 
-	return fds, nil
+	filename := fmt.Sprintf("socketpair-#%d:%d", fds[0], fds[1])
+
+	return SocketPair{
+		os.NewFile(uintptr(fds[0]), filename+"[0]"),
+		os.NewFile(uintptr(fds[1]), filename+"[1]"),
+	}, nil
 }
 
 // LocalFile returns the socketpair fd for local usage as an *os.File.
 func (fds SocketPair) LocalFile() *os.File {
-	return os.NewFile(uintptr(fds[local]), fds.fileName()+"[0]")
+	return fds.local
 }
 
 // PeerFile returns the socketpair fd for peer usage as an *os.File.
 func (fds SocketPair) PeerFile() *os.File {
-	return os.NewFile(uintptr(fds[peer]), fds.fileName()+"[1]")
+	return fds.peer
```

Comment:
Ditto here. Isn't this also changing the semantics of PeerConn() so that it can only be called once with subsequent calls failing ? Should at least be documented.